astropy/tests/plugins/config.py

```diff
@@ -0,0 +1,77 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""
+This plugin provides customization of configuration and cache directories used
+by pytest.
+"""
+import datetime
+import locale
+import os
+import sys
+from collections import OrderedDict
+
+import pytest
+
+from ...config.paths import set_temp_config, set_temp_cache
+from ...utils.argparse import writeable_directory
+from ..helper import treat_deprecations_as_exceptions
+
+import importlib.machinery as importlib_machinery
+
+
+# these pytest hooks allow us to mark tests and run the marked tests with
+# specific command line options.
+
+def pytest_addoption(parser):
+
+    parser.addoption("--config-dir", nargs='?', type=writeable_directory,
+                     help="specify directory for storing and retrieving the "
+                          "Astropy configuration during tests (default is "
+                          "to use a temporary directory created by the test "
+                          "runner); be aware that using an Astropy config "
+                          "file other than the default can cause some tests "
+                          "to fail unexpectedly")
+
+    parser.addoption("--cache-dir", nargs='?', type=writeable_directory,
+                     help="specify directory for storing and retrieving the "
+                          "Astropy cache during tests (default is "
+                          "to use a temporary directory created by the test "
+                          "runner)")
+    parser.addini("config_dir",
+                  "specify directory for storing and retrieving the "
+                  "Astropy configuration during tests (default is "
+                  "to use a temporary directory created by the test "
+                  "runner); be aware that using an Astropy config "
+                  "file other than the default can cause some tests "
+                  "to fail unexpectedly", default=None)
+
+    parser.addini("cache_dir",
+                  "specify directory for storing and retrieving the "
+                  "Astropy cache during tests (default is "
+                  "to use a temporary directory created by the test "
+                  "runner)", default=None)
+
+def pytest_configure(config):
+    treat_deprecations_as_exceptions()
+
+def pytest_runtest_setup(item):
+    config_dir = item.config.getini('config_dir')
+    cache_dir = item.config.getini('cache_dir')
+
+    # Command-line options can override, however
+    config_dir = item.config.getoption('config_dir') or config_dir
+    cache_dir = item.config.getoption('cache_dir') or cache_dir
+
+    # We can't really use context managers directly in py.test (although
+    # py.test 2.7 adds the capability), so this may look a bit hacky
```

Comment:
I'm not sure I got this perfectly. Why can't we use context managers? The pytest version requirement is already >=3.1

Reply:
The git blame shows that this comment is fairly old. I'm not sure whether this is worth fixing in this PR, but I'll take a look.

Reply:
sure, also, this one can be definitely left for a follow-up